src/doc/trpl/error-handling.md

```diff
@@ -1,308 +1,2147 @@
+simple, but to do that, we will first need to explore the `Option` and `Result`
+types. Both of these types have a method called `unwrap` defined on them.
 
-What do we mean by "recover"? Well, in most cases, the possibility of an error
-is expected. For example, consider the `parse` function:
+### The `Option` type
 
-```ignore
-"5".parse();
+The `Option` type is
+[defined in the standard library](http://doc.rust-lang.org/std/option/enum.Option.html):
+
+```rust
+enum Option<T> {
+    None,
+    Some(T),
+}
 ```
 
-This method converts a string into another type. But because it's a string, you
-can't be sure that the conversion actually works. For example, what should this
-convert to?
+The `Option` type is a way to use Rust's type system to express the
+*possibility of absence*. Encoding the possibility of absence into the type
+system is an important concept because it will cause the compiler to force the
+programmer to handle that absence. Let's take a look at an example that tries
+to find a character in a string:
 
-```ignore
-"hello5world".parse();
+```rust
+// Searches `haystack` for the Unicode character `needle`. If one is found, the
+// byte offset of the character is returned. Otherwise, `None` is returned.
+fn find(haystack: &str, needle: char) -> Option<usize> {
+    for (offset, c) in haystack.char_indices() {
+        if c == needle {
+            return Some(offset);
+        }
+    }
+    None
+}
 ```
 
-This won't work. So we know that this function will only work properly for some
-inputs. It's expected behavior. We call this kind of error a *failure*.
+Notice that when this function finds a matching character, it doen't just
+return the `offset`. Instead, it returns `Some(offset)`. `Some` is a variant or
+a *value constructor* for the `Option` type. You can think of it as a function
+with the type `fn<T>(value: T) -> Option<T>`. Correspondingly, `None` is also a
+value constructor, except it has no arguments. You can think of `None` as a
+function with the type `fn<T>() -> Option<T>`.
 
-On the other hand, sometimes, there are errors that are unexpected, or which
-we cannot recover from. A classic example is an `assert!`:
+This might seem like much ado about nothing, but this is only half of the
+story. The other half is *using* the `find` function we've written. Let's try
+to use it to find the extension in a file name.
 
 ```rust
-# let x = 5;
-assert!(x == 5);
+fn main_find() {
+    let file_name = "foobar.rs";
+    match find(file_name, '.') {
+        None => println!("No file extension found."),
+        Some(i) => println!("File extension: {}", &file_name[i+1..]),
+    }
+}
 ```
 
-We use `assert!` to declare that something is true. If it's not true, something
-is very wrong. Wrong enough that we can't continue with things in the current
-state. Another example is using the `unreachable!()` macro:
+This code uses [pattern
+matching](http://doc.rust-lang.org/1.0.0-beta.5/book/patterns.html) to do *case
```

Comment:
this link needs to be updated.

also, generally, links get moved to the `[][]` style in the book, would make this particular paragraph much nicer to read, for example